### PR TITLE
Lane C: connect GitHub accounts to a workspace, and say when a repository is not ready

### PR DESCRIPTION
**Pull request by chelojimenez**
The Inspector half of Lane C (package 7). Pairs with [MCPJam/mcpjam-backend#1037](https://github.com/MCPJam/mcpjam-backend/pull/1037), which must be on staging before this is useful — the new function ids do not exist until it deploys.

## What changes

GitHub Checks assumed **one** GitHub App installation for the whole deployment, so the settings page had nothing to say about *which* account a repository came from. The backend now binds installations to organizations; this is the surface for it.

**A "GitHub accounts" section.** Install the app on an account from here, or *claim* an installation somebody already added from GitHub's side. Claiming sends you through GitHub sign-in, and the copy says why:

> …installing the app is not on its own proof that it is yours to connect here.

That sentence is the product-facing version of the security argument in the backend PR — the App can mint a token for every installation it has, so a user-authorization leg is what proves *this person* administers *that account*.

**Explicit unbind with confirmation**, whose copy states the limit of the consequence as well as the consequence: checks stop immediately, suite and policy choices are kept, so reconnecting is not a rebuild. Copy that implied otherwise would stop admins acting on a state they need to act on.

**Connected repositories carry a state when something is wrong** — `Reconnect required`, `App inactive`, `No access` — each naming what happened and ruling out the reading that would send someone to fix the wrong thing. Three of the four states have nothing to do with the code, and the natural assumption when a check stops is that they do, so each line says "this is not a problem with your pull requests". A healthy row gets **no badge at all**: a badge saying "fine" on every row makes the ones that need attention harder to find.

**The picker aggregates every connected account and selects by GitHub's numeric repository id**, not by name — two accounts can each have a `widgets`, and a name is not an identity. The account label appears beside a repository only when it disambiguates.

**`/settings/integrations/github/callback`** handles both of GitHub's returns. The App's setup URL and its OAuth callback point at one path and are told apart by which parameters arrived.

## Three decisions worth reviewing

**Nothing is inferred client-side.** `connectionStatus` is derived by the backend from three facts the browser never sees, and in particular it is *not* inferred from a missing visibility badge — absence there means "GitHub did not tell us", not "something is wrong", and conflating them would put a warning on a healthy repository whose `private` flag GitHub happened to omit. There is a test for exactly that.

**Callback parameters are forwarded verbatim.** No parsing, no normalization, no pre-validation. The backend matches states by hash and treats `installation_id` as a claim GitHub itself documents as spoofable, so cleverness here could only turn a legitimate state into a non-matching one — while a well-meaning "clean this up first" is how a working flow breaks. The one exception is checking that `installation_id` is a positive integer, and only because there is otherwise nothing to send.

**Redirects go through one guarded helper.** `lib/github-external-redirect.ts` parses the URL and refuses anything whose origin is not `https://github.com` — not because the backend is untrusted (it builds every one of these URLs), but so this never quietly becomes a general-purpose `location.assign(userInput)` the day something else calls it. It is parsed rather than prefix-matched, because `https://github.com.evil.test/` passes a `startsWith` check and that is exactly how that ships.

## Review round 1 — and a correction

CodeRabbit caught a **real and total bug**, fixed in `191a3ef`: the callback route was registered in `ROUTE_ELEMENTS` but not in `APP_ROUTES`. `buildRouteChildren` iterates `APP_ROUTES`, so it was never mounted — every return trip from GitHub would have fallen through to the `"*"` catch-all and rendered the servers screen, making the whole binding flow unreachable.

The reason it got that far is worth more than the fix: the component tests mount the page directly in a `MemoryRouter`, so none of them exercises the route table. `buildRouteChildren` now throws on **both** directions of the invariant — it already refused a route with no element; it now also refuses an element with no route, naming the stranded path — and `route-elements-coverage.test.ts` builds the real router, so the assertion is the same check the app runs at startup. Red-probed: removing the entry again fails all three tests with the named error.

**Correction.** `191a3ef` claimed five review fixes and landed two. A failed assertion aborted the script meant to write the other three before it reached its write, and I did not re-verify the files afterwards — the suite stayed green because the tests I thought I had changed were still the old ones, passing for the old reasons. `8aeea41` actually applies them, and red-probes the two that make a behavioural claim:

- **StrictMode.** The single-use-state test used `rerender`, which keeps one component instance and so never exercised the double mount the ref exists for. Now wrapped in `StrictMode`. Probe: removing the `startedRef` guard fails it with `expected "spy" to be called 1 times, but got 2 times`; under the old version it passed.
- **Same-named repositories.** The picker-identity test used two *different* full names, so a name-keyed implementation satisfied it. Both fixtures now share one name. Probe: keying the picker on `fullName` resolves the `widgets · globex` option to `bind-acme` / `201` — the other account's repository — and fails; under the old fixture it passed.
- **Redirect-throw coverage**, which found a real defect: `UnsafeRedirectError`'s message is developer text ("Refused to redirect outside GitHub") and the setup leg's `.catch(fail)` put it on screen verbatim. A guard refusing a URL is not a backend refusal, so it now renders the flat binding copy and logs the reason for an operator.

The two that *did* land in `191a3ef`: the raw-installation-id fixture (the assertion previously could not fail, because no fixture contained an id) and the `unbound` state.

Also from that round:

- The picker's three rules were written twice, once per surface. They are a contract with the backend, so they moved to `lib/github-repo-picker.ts` — two copies drift the first time either side gains a field, and the failure mode is a connect naming the wrong installation.
- `role="status"` on the callback page's phase output — it replaces its whole content asynchronously, leaving a screen reader on a message that had already gone.
- `satisfies` on the copy maps, so a drifting status union fails to compile rather than waiting for something to index the map.
- Changeset raised to `minor`, matching how this repo versions added functionality, and it now names the backend dependency.

**Not taken:** a test pinning what the binding actions send when the organization is null. The buttons only render once availability resolves, which requires an organization, so it would pin an accident rather than a decision.

## Contract changes

`InstallationRepo` gains required `repositoryId` and optional `installationRef` / `accountLogin`; `GithubCheckRepoConfigRow` gains required `connectionStatus`. `connectVerifiedRepo` takes optional `installationRef` and `repositoryId`, both lifted straight off the picked listing entry rather than reassembled.

`installationRef` is **optional on purpose**: an organization with no binding yet is still listed through the backend's pinned installation, and those entries carry no ref. Omitting it is what keeps the compatibility connect reachable until the pin retires. Two tests cover that window.

Every one of these argument shapes is hand-mirrored — there is no generated Convex client, so a drift fails at runtime on the click. The new call shapes are commented as part of the backend's signature, and the hook's exact-set binding test is the guard that a removed id cannot be wired back in.

## Verification

- `npx vitest run --project client` — **865 files / 9325 tests passing**, 6 skipped, 0 failures, on the pushed head `8aeea41`.
- `npx tsc --noEmit -p client/tsconfig.json` — no errors in any file this PR touches.
- `npx prettier --check` on the changed files only — clean. Prettier is v2 with no config here, so no file was reformatted beyond the lines this PR changed.
- Changeset included.

Note for anyone running the suite locally: `client/src/components/evals/__tests__/suite-iterations-github-checks.test.tsx` fails to collect on a fresh checkout until `npm run sdk:build` has run (`@mcpjam/sdk/predicates` is unresolvable otherwise). Pre-existing on `main`, not introduced here.

## New tests

| | |
|---|---|
| `route-elements-coverage.test.ts` | the router mounts every route it registers — the regression above |
| `github-install-callback-route.test.tsx` | both legs, verbatim forwarding, the single-use state running once under `StrictMode`, the redirect guard throwing, the pick list, and a directly-opened callback explaining itself |
| `github-external-redirect.test.ts` | the origin allowlist, including the `github.com.evil.test` case |
| `github-checks-route.test.tsx` | the accounts section, every binding state, unbind confirm/dismiss, every connection-status badge, account-label disambiguation, and that a raw installation id present in the fixture never reaches the DOM |
| `suite-github-checks-section.test.tsx` | two identically-named repositories from different accounts resolving distinctly |
| `useGithubChecksSettings.test.tsx` | the exact binding sets for both hooks, and that the callback hook takes no organization |

---

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01FWPD216otYy9YVi1rXTGJy

### CODE REVIEW

**Comment by chelojimenez** [comment]
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scan Engine |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/marcelo-apk6cgUZZztPZRjPdWkYwH/pr-checks/d3df37b8-7a32-4d7f-8153-fa2e24e8f382?refs=8aeea4198df446bba1a4421d52626bcdaff6e137&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

File: mcpjam-inspector/client/src/router.tsx [hunk 1]
```diff
@@ -155,6 +156,14 @@ const ROUTE_ELEMENTS: Record<
...
   "settings/integrations": { element: <IntegrationsSettingsRoute /> },
   "settings/integrations/github": { element: <GithubChecksSettingsRoute /> },
+  // Where GitHub sends the browser back — BOTH the App's setup URL and its
+  // OAuth callback point here, and the page tells them apart by which query
+  // parameters arrived. One path because GitHub App settings take one of
+  // each, and because a second route would be a second place to keep the
+  // "pass everything through verbatim" rule.
+  "settings/integrations/github/callback": {
+    element: <GithubInstallCallbackSettingsRoute />,
+  },
```
[the remark this answers is not in the document]
Comment (chelojimenez) [reply mcpjam-inspector/client/src/router.tsx:175]: Confirmed and fixed in 191a3ef — this was a real bug, and a total one: every return trip from GitHub would have landed on the servers screen, so the whole binding flow was unreachable.

Added the path to `APP_ROUTES` (and to the `settings` surface's `routePatterns`, which the coverage test matches exactly).

Also closed the hole that let it through. The two directions of this invariant were asymmetric — `buildRouteChildren` already threw for an `APP_ROUTES` entry with no element, but an element with no route just vanished, and the silent one is the one that reaches a user as a wrong page. It now throws on both, naming the stranded path, and `route-elements-coverage.test.ts` builds the real router so the assertion is the same check the app runs at startup.

Red-test probe: removing the `APP_ROUTES` entry again fails all three tests with

```
[router] element registered for "settings/integrations/github/callback", which is not in APP_ROUTES — it would never be mounted
```

Worth noting why the existing tests couldn't catch this: the component tests mount the page directly in a `MemoryRouter`, so they exercise the component and never the route table.

---
_Generated by [Claude Code](https://claude.ai/code)_
